src/lem/circuit.rs

```diff
@@ -998,35 +1154,45 @@ impl Func {
                     // Now we need to enforce that at exactly one path was taken. We do that by enforcing
                     // that the sum of the previously collected `Boolean`s is one. But, of course, this
                     // irrelevant if we're on a virtual path and thus we use an implication gadget.
-                    selector.push(not_dummy.not());
+                    let dummy = not_dummy.not();
+                    selector.push(dummy);
                     enforce_selector_with_premise(
                         &mut cs.namespace(|| "enforce_selector_with_premise"),
                         not_dummy,
                         &selector,
                     )
                     .with_context(|| " couldn't constrain `enforce_selector_with_premise`")
                 }
-                Ctrl::MatchVal(match_var, cases, def) => {
-                    let match_lit = bound_allocations.get(match_var)?.hash().clone();
+                Ctrl::Match(match_var, cases, def) => {
+                    let match_lit = bound_allocations.get(match_var)?.clone();
                     let mut selector = Vec::with_capacity(cases.len() + 2);
                     let mut branch_slots = Vec::with_capacity(cases.len());
                     for (i, (lit, block)) in cases.iter().enumerate() {
-                        let lit_ptr = lit.to_ptr(g.store);
+                        let lit_ptr = lit.to_ptr_cached(g.store);
+                        let lit_tag = g.store.hash_ptr(&lit_ptr)?.tag.to_field::<F>();
                         let lit_hash = g.store.hash_ptr(&lit_ptr)?.hash;
```

Comment:
Is `hash_ptr` so cheap that we don't want to just do it once and then extract the tag and hash from the result?

Reply:
It is indeed cheap at this point (after hydration), but that's something that needs to be coded better. I will change it